modules/openapi-generator/src/main/resources/Java/libraries/apache/ApiClient.mustache

```diff
@@ -0,0 +1,1039 @@
+          url.append(escapeString(param.getName())).append("=").append(value);
+        }
+      }
+    }
+
+    return url.toString();
+  }
+
+  protected boolean isSuccessfulStatus(int statusCode) {
+    return statusCode >= 200 && statusCode < 300;
+  }
+
+  protected boolean isBodyAllowed(String method) {
+    return bodyMethods.contains(method);
+  }
+
+  protected Cookie buildCookie(String key, String value, URI uri) {
+    BasicClientCookie cookie = new BasicClientCookie(key, value);
+    cookie.setDomain(uri.getHost());
+    cookie.setPath("/");
+    return cookie;
+  }
+
+  protected <T> T processResponse(CloseableHttpResponse response, TypeReference<T> returnType) throws ApiException, IOException {
+    statusCode = response.getStatusLine().getStatusCode();
+    if (statusCode == HttpStatus.SC_NO_CONTENT) {
+      return null;
+    }
+
+    responseHeaders = transformResponseHeaders(response.getAllHeaders());
+    if (isSuccessfulStatus(statusCode)) {
+      return this.deserialize(response, returnType);
+    } else {
+      String message = EntityUtils.toString(response.getEntity());
+      throw new ApiException(message, statusCode, responseHeaders, message);
+    }
+  }
+
+  /**
+   * Invoke API by sending HTTP request with the given options.
+   *
+   * @param <T> Type
+   * @param path The sub-path of the HTTP URL
+   * @param method The request method, one of "GET", "POST", "PUT", and "DELETE"
+   * @param queryParams The query parameters
+   * @param collectionQueryParams The collection query parameters
+   * @param body The request body object - if it is not binary, otherwise null
+   * @param headerParams The header parameters
+   * @param cookieParams The cookie parameters
+   * @param formParams The form parameters
+   * @param accept The request's Accept header
+   * @param contentType The request's Content-Type header
+   * @param authNames The authentications to apply
+   * @return The response body in type of string
+   * @throws ApiException API exception
+   */
+   public <T> T invokeAPI(
+       String path,
+       String method,
+       List<Pair> queryParams,
+       List<Pair> collectionQueryParams,
+       Object body,
+       Map<String, String> headerParams,
+       Map<String, String> cookieParams,
+       Map<String, Object> formParams,
+       String accept,
+       String contentType,
+       String[] authNames,
+       TypeReference<T> returnType) throws ApiException {
+    if (body != null && !formParams.isEmpty()) {
+      throw new ApiException("Cannot have body and form params");
+    }
+
+    updateParamsForAuth(authNames, queryParams, headerParams, cookieParams);
+    final String url = buildUrl(path, queryParams, collectionQueryParams);
+
+    RequestBuilder builder = RequestBuilder.create(method);
+    builder.setUri(url);
+
+    RequestConfig config = RequestConfig.custom()
+      .setConnectionRequestTimeout(connectionTimeout)
```

Comment:
I could also set the socket timeout here, not sure if that would be expected behaviour though